pkg/controller/aci_istio.go

```diff
@@ -0,0 +1,294 @@
+
+	informer := cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				return aciistioclient.AciV1().AciIstioOperators(os.Getenv("SYSTEM_NAMESPACE")).List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return aciistioclient.AciV1().AciIstioOperators(os.Getenv("SYSTEM_NAMESPACE")).Watch(options)
+			},
+		},
+		&istiooperators.AciIstioOperator{},
+		0,
+		cache.Indexers{},
+	)
+
+	controller := &AciIstioController{
+		Logger:    log.NewEntry(log.New()),
+		Clientset: aciistioclient,
+		Informer:  informer,
+		Queue:     queue,
+		Handlers:  &AciIstioOperatorHandler{},
+	}
+
+	log.Info("Adding Event Handlers")
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			log.Infof("The aci-istio-operator key: %s", key)
+			if err == nil {
+				queue.Add(key)
+			}
+		},
+		UpdateFunc: func(prevObj, currentObj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(currentObj)
+			log.Infof("Updated aci-istio-operator: %s", key)
+			if err == nil {
+				queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			log.Infof("Deleted aci-istio-operator key is : %s", key)
+			if err == nil {
+				queue.Add(key)
+			}
+		},
+	})
+	return controller
+}
+
+func (c *AciIstioController) Run(stopCh <-chan struct{}) {
+	// Handling Panic gracefully
+	defer utilruntime.HandleCrash()
+	//Shutdown queue when go routine is done for the queue
+	defer c.Queue.ShutDown()
+
+	c.Logger.Info("Controller.Run: initiating")
+
+	// Run informer to start watching and listening
+	go c.Informer.Run(stopCh)
+
+	// Sync the current resources
+	if !cache.WaitForCacheSync(stopCh, c.HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("Controller.Sync: Error syncing the cache"))
+		return
+	}
+	c.Logger.Info("Controller.Sync: Cache sync complete")
+
+	// Run worker in 1 sec interval
+	wait.Until(c.runWorker, time.Second, stopCh)
+}
+
+func (c *AciIstioController) HasSynced() bool {
+	return c.Informer.HasSynced()
+}
+
+// runWorker executes new objects added to queue in loop
+func (c *AciIstioController) runWorker() {
+	log.Info("Controller.runWorker: starting the controller")
+
+	for c.processNextItem() {
+		log.Info("Controller.runWorker: processing the next item")
+	}
+
+	log.Info("Controller.runWorker: completed")
+}
+
+func (c *AciIstioController) processNextItem() bool {
+	log.Info("Controller.processNextItem: starting....")
+
+	key, quit := c.Queue.Get()
+
+	if quit {
+		return false
+	}
+
+	defer c.Queue.Done(key)
+
+	keyRaw := key.(string)
+
+	item, exists, err := c.Informer.GetIndexer().GetByKey(keyRaw)
+	if err != nil {
+		if c.Queue.NumRequeues(key) < 10 {
+			c.Logger.Errorf("Controller.processNextItem: Failed to process the item with key %s with error %v, retrying", key, err)
```

Comment:
nit: I am not sure we want to mention "Controller.processNextItem" since i believe the logger will add, however if we do want to perhaps change the name to AciIstioController